dask/array/random.py

```diff
@@ -17,7 +17,595 @@
+        self._bit_generator = bit_generator
+        self._generator = Generator
+
+    def __str__(self):
+        _str = self.__class__.__name__
+        _str += "(" + self._bit_generator.__class__.__name__ + ")"
+        return _str
+
+    def _wrap(
+        self, funcname, *args, size=None, chunks="auto", extra_chunks=(), **kwargs
+    ):
+        """Wrap numpy random function to produce dask.array random function
+
+        extra_chunks should be a chunks tuple to append to the end of chunks
+        """
+        if size is not None and not isinstance(size, (tuple, list)):
+            size = (size,)
+
+        shapes = list(
+            {
+                ar.shape
+                for ar in chain(args, kwargs.values())
+                if isinstance(ar, (Array, np.ndarray))
+            }
+        )
+        if size is not None:
+            shapes.append(size)
+        # broadcast to the final size(shape)
+        size = broadcast_shapes(*shapes)
+        chunks = normalize_chunks(
+            chunks,
+            size,  # ideally would use dtype here
+            dtype=kwargs.get("dtype", np.float64),
+        )
+        slices = slices_from_chunks(chunks)
+
+        def _broadcast_any(ar, shape, chunks):
+            if isinstance(ar, Array):
+                return broadcast_to(ar, shape).rechunk(chunks)
+            if isinstance(ar, np.ndarray):
+                return np.ascontiguousarray(np.broadcast_to(ar, shape))
+
+        # Broadcast all arguments, get tiny versions as well
+        # Start adding the relevant bits to the graph
+        dsk = {}
+        lookup = {}
+        small_args = []
+        dependencies = []
+        for i, ar in enumerate(args):
+            if isinstance(ar, (np.ndarray, Array)):
+                res = _broadcast_any(ar, size, chunks)
+                if isinstance(res, Array):
+                    dependencies.append(res)
+                    lookup[i] = res.name
+                elif isinstance(res, np.ndarray):
+                    name = f"array-{tokenize(res)}"
+                    lookup[i] = name
+                    dsk[name] = res
+                small_args.append(ar[tuple(0 for _ in ar.shape)])
+            else:
+                small_args.append(ar)
+
+        small_kwargs = {}
+        for key, ar in kwargs.items():
+            if isinstance(ar, (np.ndarray, Array)):
+                res = _broadcast_any(ar, size, chunks)
+                if isinstance(res, Array):
+                    dependencies.append(res)
+                    lookup[key] = res.name
+                elif isinstance(res, np.ndarray):
+                    name = f"array-{tokenize(res)}"
+                    lookup[key] = name
+                    dsk[name] = res
+                small_kwargs[key] = ar[tuple(0 for _ in ar.shape)]
+            else:
+                small_kwargs[key] = ar
+
+        sizes = list(product(*chunks))
+        seeds = random_rng_data(len(sizes), self._bit_generator.state)
```

Comment:
Hi, everyone! This PR was linked from a [discussion on `data-apis/array-api`](https://github.com/data-apis/array-api/issues/431) about standardizing PRNG interfaces.

I haven't fully grokked the whole scheme here, but I would recommend against this method of deriving parallel PRNG states for the distributed computation. Using the output of a PRNG to reseed itself typically creates [a random non-invertible mapping](https://www.pcg-random.org/posts/too-big-to-fail.html) which has poor statistical behavior. Some PRNGs are designed and tested for this, but the `BitGenerator`s currently in numpy have not been designed with this usage in mind. Even ones that are designed and tested for this don't compress the state down to a bottleneck of 64-bit integers. You are bound to have a Birthday Collision in only a few billion calls (which sounds like a lot, but is far below the safety margin these PRNGs are designed to provide otherwise).

This particular invocation is problematic because you aren't advancing `self._bit_generator` but working from a copy.

Since you are working with `BitGenerator`s, they do carry around the `SeedSequence` they were instantiated with. Spawning from the `SeedSequence` is a much safer and statistically sound method of generating new states to pass to parallel processes. We were conservative in the initial release, so we did not make it public-public, but it is reliably available here as `self._bit_generator._seed_seq`. For example, I hereby bless the use of this function to spawn a requested number of `BitGenerator`s from the given one.

```python
def spawn_bitgens(bitgen, n_bitgens):
    seeds = bitgen._seed_seq.spawn(n_bitgens)
    bitgens = [type(bitgen)(seed) for seed in seeds]
    return bitgens
```

We will be [lifting up the spawning interface](https://github.com/data-apis/array-api/issues/431#issuecomment-1144054913) to the `BitGenerator` and `Generator` soon, I think. It would make it a lot easier for everyone. But we'll keep that `._seed_seq` attribute around.

FWIW, you can implement a `dask.base.normalize_bit_generator()` like so:

```python
@normalize_token.register(np.random.BitGenerator)
def normalize_bit_generator(bg):
    return normalize_token(bg.state)
```

Reply:
Thanks for taking the time to look at this Robert! 🙏

We do reseed [here]( https://github.com/dask/dask/blob/f075e86fd4f68ede1ee989ef9488955b0d653884/dask/multiprocessing.py#L252-L256 ). Though Idk if that needs to be updated for the new API or would work with both. Admittedly there may be other changes needed in this PR as well

Reply:
That's separate from what I'm talking about here. When I talk about "using the output of a PRNG to reseed itself", I'm referring to using the output `seeds` here in `_apply_random_func()` to instantiate new `Generator` instances.

Reply:
Thought that might be the case

Though did want to surface that `seed`ing behavior as it might not be immediately obvious Dask is doing that. Also couldn't tell if we needed to change it as part of this work

Reply:
No, that's generally a good idea to avoid [The Mistake](https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects/).

Reply:
Thanks a lot for your comments and suggestions. Committed via 4d561f64.